### DISCUSSION TITLE
Usage with DB data in a volume?

### DISCUSSION THREAD

**Discussion by jpbaril** [pgautoupgrade/docker-pgautoupgrade#54] [Q&A] [answered]
How to use this tool with database data that is stored in a Docker volume and not a bind directory ?

**Comment by justinclift** [comment]
No idea (yet).  If the database directory is available inside the container at some path (ie `/foo/bar`), then in theory things should work anyway if your `PGDATA` environment variable is set to that path.  Along the lines of:

```
$ docker compose -e PGDATA=/foo/bar pgautoupgrade/pgautoupgrade:latest
```

Does that help? :smile:

**Reply by jpbaril** [reply]
> If the database directory is available inside the container at some path

That's indeed where the challenge lies.

I tried to map it directly to the Docker volume in /var/lib/docker/volumes/.

It seemed to work at first but I now face another issue. I get this:
"connection to server on socket "/var/run/postgresql/.s.PGSQL.50432" failed: FATAL:  role "postgres" does not exist"

FYI, I think my DB has instead a role that is the same as the app name.

**Reply by justinclift** [reply]
No worries.  As a potentially useful data point, if you're familiar with shell scripting then the pgautoupgrade container is 99% the standard Docker PostgreSQL container with some shell scripting added in the right strategic spot:

https://github.com/pgautoupgrade/docker-pgautoupgrade/blob/e8a81eeb2232b0584e1ce85d1ca716977023a52f/docker-entrypoint.sh#L371-L585

Mentioning that because there are several other environment variables used there (we should document those @spwoodcock?).  The one you're probably after is the `POSTGRES_USER` variable, used on (at least) line 544.

So, give that a go and see if things work? :smile:

**Reply by jpbaril** [reply]
Yeah, I did search for such environmental variable but it looks like I did not look carefully enough. 

Still, it did not work.
Now I have this error message in log file "/var/lib/postgresql/data/new/pg_upgrade_output.d/20241006T045617.218/log/pg_upgrade_utility.log":
```
vacuumdb: vacuuming database "postgres"
vacuumdb: error: processing of database "postgres" failed: ERROR:  unexpected data beyond EOF in block 19 of relation base/5/2619
HINT:  This has been seen to occur with buggy kernels; consider updating your system.
```

**Reply by justinclift** [reply]
Uh, that's not good.  That seems to be saying the on disk format for your database is corrupted. :frowning:

---

Hmmm, we don't yet have a way of skipping the (recently added) background tasks that automatically run after the main part of the upgrade process.  Those background tasks are things like `vacuum analyze` and database reindexing.

What you *could* do, is start your docker compose with the `PGAUTO_DEVEL=server` environment variable (in addition to the others).

That's meant for development, but will run the main upgrade process then pause before starting the further background tasks or the main PostgreSQL process.  So, you could start the container with the above argument (and keep an eye on the logs), then when the upgrade is done you tell docker to shut down the container.

In theory that should result in a database with upgraded data files, but without `vacuum` (etc) having been run on the result.  You'd want to manually check if the database is ok at that point (ie does it start, etc), and you'd probably also want to then manually run `vacuum analyze` and `index database` on all of the databases.

**Reply by jpbaril** [reply]
> That seems to be saying the on disk format for your database is corrupted. 😦

ok. So I just retried with the backuped original data and now with the use of the required environmental variable it works beautifully. Thanks.

**Comment by spwoodcock** [comment]
Using a docker volume instead of a bind mount should be as simple as:

`--volume name-of-volume:/var/lib/postgresql/data`

Or if using compose:

```yaml
volumes:
  db-data-volume:
services:
  upgrade:
    volumes:
      - db-data-volume:/var/lib/postgresql/data
```

---

Agree @justinclift it would be great to have env vars / config options documented on a wiki page. I'm currently on holidays, but back in a week & will hopefully have some time for writing up some docs 👍

**Reply by jpbaril** [reply]
> Using a docker volume instead of a bind mount should be as simple as:
> 
> `--volume name-of-volume:/var/lib/postgresql/data`
> 
> Or if using compose:
> 
> ```yaml
> volumes:
>   db-data-volume:
> services:
>   upgrade:
>     volumes:
>       - db-data-volume:/var/lib/postgresql/data
> ```

In the end as I had mentioned I just used direct path to data in  /var/lib/docker/volumes/ even if not ideal.
Yet, for the compose version, I'm not sure and did not test it but would it not also need "external: true" since the volume would have already been created by another container ?

**Reply by spwoodcock** [reply, marked as answer]
Using the docker volume directory for bind mounts is not advised, as it could cause permission issues or even corruption.

But yes you are right, if the docker volume is already created you should reference the volume like this:

```yaml
volumes:
  db-data-volume:
    external: true
    name: the-name-of-the-existing-volume
services:
  upgrade:
    volumes:
      - db-data-volume:/var/lib/postgresql/data
```
